File: backend_node/src/services/admin.service.js
const { pool } = require('../db')
const encryption = require('../encryption')
const { toBool, formatDateTime } = require('../mappers')
const { ApiError } = require('../httpError')

const DECLARATION_SELECT = `
  SELECT d.*, u.prenom AS clientPrenom, u.nom AS clientNom, u.email AS clientEmail
  FROM declarations d
  LEFT JOIN utilisateurs u ON u.id = d.client_id
`

function declToDto(r) {
  return {
    id: r.id,
    anneeFiscale: r.annee_fiscale,
    statut: r.statut,
    dateSoumission: formatDateTime(r.date_soumission),
    avecConjoint: toBool(r.avec_conjoint),
    donneesFormulaire: encryption.decrypt(r.donnees_formulaire),
    clientId: r.client_id,
    clientNom: r.clientPrenom && r.clientNom ? `${r.clientPrenom} ${r.clientNom}` : null,
    clientEmail: r.clientEmail,
  }
}

async function getStats() {
  const [[{ c: totalUtilisateurs }]] = await pool.query('SELECT COUNT(*) AS c FROM utilisateurs')
  const [[{ c: totalDeclarations }]] = await pool.query('SELECT COUNT(*) AS c FROM declarations')
  const [[{ c: enAttente }]] = await pool.query("SELECT COUNT(*) AS c FROM declarations WHERE statut = 'EN_ATTENTE'")
  const [[{ c: validees }]] = await pool.query("SELECT COUNT(*) AS c FROM declarations WHERE statut = 'VALIDEE'")
  const [[{ c: rejetees }]] = await pool.query("SELECT COUNT(*) AS c FROM declarations WHERE statut = 'REJETEE'")
  return { totalUtilisateurs, totalDeclarations, enAttente, validees, rejetees }
}

async function getAllUsers() {
  const [rows] = await pool.query(
    'SELECT id, email, prenom, nom, role, date_creation AS dateCreation FROM utilisateurs ORDER BY date_creation ASC'
  )
  return rows.map((r) => ({
    id: r.id,
    email: r.email,
    prenom: r.prenom,
    nom: r.nom,
    role: r.role,
    dateCreation: formatDateTime(r.dateCreation),
  }))
}

async function getAllDeclarations() {
  const [rows] = await pool.query(`${DECLARATION_SELECT} ORDER BY d.id ASC`)
  return rows.map(declToDto)
}

async function setStatut(id, statut) {
  const [rows] = await pool.query(`${DECLARATION_SELECT} WHERE d.id = ?`, [id])
  if (!rows[0]) throw new ApiError(404, 'Declaration non trouvee')
  await pool.query('UPDATE declarations SET statut = ? WHERE id = ?', [statut, id])
  const [updated] = await pool.query(`${DECLARATION_SELECT} WHERE d.id = ?`, [id])
  return declToDto(updated[0])
}

module.exports = { getStats, getAllUsers, getAllDeclarations, setStatut }

File: backend/src/main/java/com/impotfacile/service/AdminService.java
package com.impotfacile.service;

import com.impotfacile.dto.AdminDeclarationDto;
import com.impotfacile.dto.AdminUserDto;
import com.impotfacile.exception.ApiException;
import com.impotfacile.model.identity.Client;
import com.impotfacile.model.declaration.Declaration;
import com.impotfacile.model.declaration.StatutDeclaration;
import com.impotfacile.repository.*;
import com.impotfacile.security.EncryptionService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AdminService {

    private final UtilisateurRepository utilisateurRepository;
    private final DeclarationRepository declarationRepository;
    private final EncryptionService encryptionService;

    public Map<String, Object> getStats() {
        return Map.of(
                "totalUtilisateurs", utilisateurRepository.count(),
                "totalDeclarations", declarationRepository.count(),
                "enAttente", declarationRepository.countByStatut(StatutDeclaration.EN_ATTENTE),
                "validees", declarationRepository.countByStatut(StatutDeclaration.VALIDEE),
                "rejetees", declarationRepository.countByStatut(StatutDeclaration.REJETEE)
        );
    }

    public List<AdminUserDto> getAllUsers() {
        return utilisateurRepository.findAll().stream()
                .map(u -> AdminUserDto.builder()
                        .id(u.getId())
                        .email(u.getEmail())
                        .prenom(u.getPrenom())
                        .nom(u.getNom())
                        .role(u.getRole())
                        .dateCreation(u.getDateCreation())
                        .build())
                .toList();
    }

    public List<AdminDeclarationDto> getAllDeclarations() {
        return declarationRepository.findAll().stream()
                .map(this::toDto)
                .toList();
    }

    @Transactional
    public AdminDeclarationDto validateDeclaration(String id) {
        Declaration decl = declarationRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Declaration non trouvee"));
        decl.setStatut(StatutDeclaration.VALIDEE);
        declarationRepository.save(decl);
        return toDto(decl);
    }

    @Transactional
    public AdminDeclarationDto rejectDeclaration(String id) {
        Declaration decl = declarationRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Declaration non trouvee"));
        decl.setStatut(StatutDeclaration.REJETEE);
        declarationRepository.save(decl);
        return toDto(decl);
    }

    private AdminDeclarationDto toDto(Declaration d) {
        Client client = d.getClient();
        return AdminDeclarationDto.builder()
                .id(d.getId())
                .anneeFiscale(d.getAnneeFiscale())
                .statut(d.getStatut())
                .dateSoumission(d.getDateSoumission())
                .avecConjoint(d.getAvecConjoint())
                .donneesFormulaire(encryptionService.decrypt(d.getDonneesFormulaire()))
                .clientId(client != null ? client.getId() : null)
                .clientNom(client != null ? client.getPrenom() + " " + client.getNom() : null)
                .clientEmail(client != null ? client.getEmail() : null)
                .build();
    }
}
